graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py

```diff
@@ -0,0 +1,1560 @@
+
+        As an additional side-effect the function populates the values of
+        `self.node_chunk_counts` for each edge type.
+
+        Parameters
+        ----------
+        node_configs : Sequence[NodeConfig]
+            A list of `NodeConfig` objects that contain the information for each node type
+            to be processed.
+
+        Returns
+        -------
+        Dict
+            A dict entry for the node_data key of the output metadata.json.
+        """
+        node_data_dict = {}  # type: Dict[str, Dict]
+        self.graph_info["nfeat_size"] = {}
+        self.graph_info["ntype_label"] = []
+        self.graph_info["ntype_label_property"] = []
+        for node_config in node_configs:
+            logging.info("node_config: %s", node_config)
+            files = node_config.files
+            file_paths = [f"{self.input_prefix}/{f}" for f in files]
+
+            separator = node_config.separator
+            node_type = node_config.ntype
+            node_col = node_config.node_col
+
+            read_nodefile_start = perf_counter()
+            # TODO: Maybe we use same enforced type for Parquet and CSV
+            # to ensure consistent behavior downstream?
+            if node_config.format == "csv":
+                node_file_schema = schema_utils.parse_node_file_schema(node_config)
+                nodes_df_untyped = self.spark.read.csv(path=file_paths, sep=separator, header=True)
+                nodes_df_untyped = nodes_df_untyped.select(node_file_schema.fieldNames())
+                # Select only the columns referenced in the config
+                # and cast each column to the correct type
+                nodes_df = nodes_df_untyped.select(
+                    *(
+                        F.col(col_name).cast(col_field.dataType).alias(col_name)
+                        for col_name, col_field in zip(
+                            node_file_schema.fieldNames(), node_file_schema.fields
+                        )
+                    )
+                )
+            else:
+                nodes_df = self.spark.read.parquet(*file_paths)
+
+            # TODO: Assert columns from config exist in the nodes df
+            self.timers["node_read_file"] += perf_counter() - read_nodefile_start
+
+            node_id_map_start_time = perf_counter()
+            if node_type in self.node_mapping_paths:
+                logging.warning(
+                    "Encountered node type '%s' that already has "
+                    "a mapping, skipping map creation",
+                    node_type,
+                )
+                # Use the existing mapping to convert the ids from str-to-int
+                mapping_df = self.spark.read.parquet(
+                    *[
+                        f"{self.output_prefix}/{map_path}"
+                        for map_path in self.node_mapping_paths[node_type]
+                    ]
+                )
+                node_df_with_ids = mapping_df.join(
```

Comment:
We have a condition `enable_assertions` that triggers that kind of check only when enabled in code, because these trigger computations potentially adversely affecting performance. We do this for the edges, so let me do the same here for nodes.